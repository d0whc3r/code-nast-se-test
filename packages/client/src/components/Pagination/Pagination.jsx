import { usePagination } from '../../context/Pagination.context'
import './Pagination.styles.css'

function Pagination() {
  const { state, dispatch } = usePagination()

  const isFirst = state.page === 1
  const isLast = state.page === state.totalPages

  const handlePrevious = () => {
    dispatch({ type: 'PREVIOUS' })
  }

  const handleNext = () => {
    dispatch({ type: 'NEXT' })
  }

  const handleFirst = () => {
    dispatch({ type: 'FIRST' })
  }

  const handleLast = () => {
    dispatch({ type: 'LAST' })
  }

  return (
    <div className="container">
      <button disabled={isFirst} className="button-text" onClick={handleFirst}>First</button>
      <button disabled={isFirst} className="button-text" onClick={handlePrevious}>Previous</button>
      <button disabled={isLast} className="button-text" onClick={handleNext}>Next</button>
      <button disabled={isLast} className="button-text" onClick={handleLast}>Last</button>
    </div>
  )
}

export default Pagination
