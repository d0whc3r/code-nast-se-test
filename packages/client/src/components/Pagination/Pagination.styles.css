.container {
    @apply mt-5 flex flex-row items-center justify-center gap-4;
}
