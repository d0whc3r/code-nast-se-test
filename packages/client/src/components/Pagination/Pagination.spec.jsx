import { describe, expect, it } from 'vitest'
import { act, fireEvent, renderWithPagination, screen } from '../../../test/utils'
import Pagination from './Pagination'

describe('Pagination', () => {
  const initialPagination = { page: 1, totalPages: 10 }
  it('should render', () => {
    const { baseElement } = renderWithPagination(<Pagination />, { initialPagination })
    expect(baseElement).toMatchSnapshot()
  })
  it('should first button must be disabled', () => {
    renderWithPagination(<Pagination />, { initialPagination })
    expect(screen.getByText(/First/)).toHaveAttribute('disabled')
  })
  it('should previous button must be disabled', () => {
    renderWithPagination(<Pagination />, { initialPagination })
    expect(screen.getByText(/Previous/)).toHaveAttribute('disabled')
  })
  it('should last button must be disabled', () => {
    renderWithPagination(<Pagination />, { initialPagination: { ...initialPagination, page: 10 } })
    expect(screen.getByText(/Last/)).toHaveAttribute('disabled')
  })
  it('should next button must be disabled', () => {
    renderWithPagination(<Pagination />, { initialPagination: { ...initialPagination, page: 10 } })
    expect(screen.getByText(/Next/)).toHaveAttribute('disabled')
  })
  it('should click first button', async () => {
    renderWithPagination(<Pagination />, { initialPagination: { ...initialPagination, page: 2 } })
    expect(screen.getByText(/First/)).not.toHaveAttribute('disabled')
    await act(() => {
      fireEvent.click(screen.getByText(/First/))
    })
    expect(screen.getByText(/First/)).toHaveAttribute('disabled')
  })
  it('should click next button', async () => {
    renderWithPagination(<Pagination />, { initialPagination: { ...initialPagination, page: 9 } })
    expect(screen.getByText(/Next/)).not.toHaveAttribute('disabled')
    await act(() => {
      fireEvent.click(screen.getByText(/Next/))
    })
    expect(screen.getByText(/Next/)).toHaveAttribute('disabled')
    expect(screen.getByText(/Last/)).toHaveAttribute('disabled')
  })
  it('should click previous button', async () => {
    renderWithPagination(<Pagination />, { initialPagination: { ...initialPagination, page: 2 } })
    expect(screen.getByText(/Previous/)).not.toHaveAttribute('disabled')
    await act(() => {
      fireEvent.click(screen.getByText(/Previous/))
    })
    expect(screen.getByText(/Previous/)).toHaveAttribute('disabled')
    expect(screen.getByText(/First/)).toHaveAttribute('disabled')
  })
  it('should click last button', async () => {
    renderWithPagination(<Pagination />, { initialPagination: { ...initialPagination, page: 2 } })
    expect(screen.getByText(/Last/)).not.toHaveAttribute('disabled')
    await act(() => {
      fireEvent.click(screen.getByText(/Last/))
    })
    expect(screen.getByText(/Last/)).toHaveAttribute('disabled')
    expect(screen.getByText(/Next/)).toHaveAttribute('disabled')
  })
})
