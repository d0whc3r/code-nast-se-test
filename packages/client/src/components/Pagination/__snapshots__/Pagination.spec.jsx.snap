// Vitest Snapshot v1

exports[`Pagination > should render 1`] = `
<body>
  <div>
    <div
      class="container"
    >
      <button
        class="button-text"
        disabled=""
      >
        First
      </button>
      <button
        class="button-text"
        disabled=""
      >
        Previous
      </button>
      <button
        class="button-text"
      >
        Next
      </button>
      <button
        class="button-text"
      >
        Last
      </button>
    </div>
  </div>
</body>
`;
